proposals/1125_dynamic-supervisor-oidc-clients/README.md

```diff
@@ -0,0 +1,495 @@
+against the Kubernetes APIs of all clusters associated with the FederationDomain.
+
+It seems desirable for an admin to explicitly choose which clients are associated with a FederationDomain. For example,
+if an admin has a FederationDomain for all the Kubernetes clusters used by the finance department, and another
+FederationDomain for all the clusters used by the HR department, then a webapp for the finance department developers
+should not necessarily be allowed to perform actions on the Kubernetes API of the HR department's clusters.
+
+###### Option 1: Explicitly associate specific clients with issuers
+
+Each FederationDomain could list which clients are allowed. For example:
+
+```yaml
+apiVersion: config.supervisor.pinniped.dev/v1alpha1
+kind: FederationDomain
+metadata:
+  namespace: pinniped-supervisor
+  name: my-domain
+spec:
+  issuer: https://my-issuer.pinniped.dev/issuer
+  # This is the new part...
+  clientRefs:
+    - "my-webapp-client"
+    - "my-other-webapp-client"
+```
+
+The `pinniped-cli` client does not need to be listed, since it only makes sense to allow `kubectl` access to all users
+of all FederationDomains. Additionally, the `pinniped-cli` can only redirect authcodes to localhost listeners,
+effectively only allowing users to log into their own accounts on their own computers.
+
+###### Option 2: Implicitly associate all clients with all issuers
+
+Rather than explicitly listing which clients are allowed on a FederationDomain, all FederationDomains could assume that
+all clients are available for use.
+
+Advantages:
+
+- Slightly less configuration for the user.
+- Slightly less implementation effort since the FederationDomain watching controller would not need to change to read
+  the list of `clientRefs`.
+
+Disadvantages:
+
+- Reusing the example scenario from above (finance and HR clusters), there would be no way to prevent a webapp for
+  finance developer users from performing operations against the clusters of HR developer users who log into the finance
+  webapp. This could be a serious security problem after the planned multiple identity providers feature is implemented
+  to allow for more differentiation between the users of the two FederationDomains. This problem is compounded by the
+  fact that many upstream OIDC providers use browser cookies to avoiding asking an active user to interactively log in
+  again, and also by the fact that we decided to punt implementing a user consent UI screen. Together, these imply that
+  an attacker from the finance department cluster which runs the client webapp would only need to trick an HR user into
+  clicking on a single link in their web browser or email client for the attacker to be able to gain access to the HR
+  clusters using the identity of the HR user, with no further interaction required by the HR user beyond just clicking
+  on the link.
+
```

Comment:
Isn't there a third option where the association is configured on the `OIDCClient`? Since only admins can configure clients, this should be as safe as configuring it on the `FederationDomain`. It  has the advantages of not requiring a `FederationDomain` update when you create a new client, and not requiring implementation work on our part to change the federation domain controller.

Reply:
Yes, that would also be possible.

The `FederationDomain` is the configuration of the downstream OIDC issuer, so in my mind it would make sense for the issuer to decide for itself which clients are allowed on the issuer. The burden on the user is about the same, i.e. when you create or delete a new issuer then you don't need to update any `OIDCClients` configuration. I'd expect us to do the same thing when we implement multiple IDPs, which is list the allowed IDPs on the `FederationDomain`, so that would be consistent if we use the same approach here. (In the case of IDPs it would also allow us to put extra per-`FederationDomain` configuration on each IDP at that "join" point, but I don't think that applies here for `OIDCClients`.)

But either alternative would be possible as an API.